### ISSUE TITLE
Multiple same type widgets on the same dashboard - what is the limitation?

### ISSUE THREAD

**Issue by dardanbekteshi** [Hygieia/Hygieia#2481]
Hi, 

I am trying to implement multiple same type widgets on a single dashboard. For example, having multiple GitLab repositories or multiple Jenkins builds in the same dashboard. 

Looking at the MongoDB and Java classes, I can see that each dashboard contains an application,  which contains an array of components.  A component then contains a map of <CollectorType, List<CollectorItem>>, meaning that in theory, it can contain multiple Gitlab repos or Jenkins builds. Now it looks as if the current implementation should in theory allow multiple same type widgets, but that is not the case. I can see two ways of doing this. For example, let's say we want two GitLab repositories on the same dashboard: 

1) A dashboard contains an application, and this application contains two components. 
      Component A: has GitLab repo 1
      Component B: has GitLab repo 2
2) A dashboard contains an application, the application contains one component.
     Component A: has two GitLab repos 

Since I can't really wrap up my head around this, could you explain why currently multiple same type widgets are not supported, and what is the limitation? Just FYI, I am not a front-end guy, so I haven't investigated the UI that much, maybe the limitation is on that side?

Thanks

**Comment by dardanbekteshi** [comment]
An answer I got on Gitter from @nireeshT, just in case someone else is interested into this question: 

> "your 2nd option is feasible way for implementing multiple same type widget. Consider widget as collector_item in component. Component is dashboard"

**Comment by HweeWong** [comment]
Hi @dardanbekteshi, I am looking for a solution for this as well. Do you think there is any backend code changes required? Or is it just UI changes?

**Comment by dardanbekteshi** [comment]
Hi @HweeWong, I am not entirely sure, and that's why I raised this issue. But from what I have seen, changes are required in both backend and frontend.

I just read (#2495) that you have already done some UI changes, while I have done some research on the backend side. I have created the following diagram which consists of the mongoDB and Java code mapping.
 
![backend-diagram](https://user-images.githubusercontent.com/23189173/46967014-92781e80-d0af-11e8-9849-5023bb8e29fa.png)

I think that we should join forces regarding this feature, where I can focus on the backend and you on the frontend? If you agree, then I suggest that you create a branch (e.g. multiple-widgets) and push the UI related code. I'll then have a look at it on Friday and start to make the backend changes accordingly. At the moment, I only work once a week on Hygieia since I am involved in another project the other four days, but I think we can still have progress if we work together :)

**Comment by stevegal** [comment]
Doesn’t the config for a widget contain an array of collector items? The collector items are gathered at the backend and only linked through the widget config to the collector item... at least that’s what I thought when I recently did the log widget. And I too would appreciate multiple widgets!

**Comment by stevegal** [comment]
From what I can see from looking at the `widgetConfigManager.js` and `widgetConfigManager.html` it adds/removes widgets based on widgetTitle. I reckon if you could change that and just expose an addWidget and removeWidget that checks for all widgets that start with the title and simply adds with a new title then it may work. Just not sure if the widget type is based on the title.. If so that'd have to change as well.

**Comment by divashdr** [comment]
Would love to have this feature available to us as soon as earliest. Thanks!

**Comment by HweeWong** [comment]
@dardanbekteshi For the extra widget, I have just duplicated the repo widget under UI components to see how it works - ideally we wouldn't want to do this manually.. I am not a frontend person as well. Thanks for posting the diagram, it does make it much clearer for me to understand what is going on.

Am I right saying that the widget is linked to a dashboard and application, hence not allowing us to show multiple widget?  (componentId : modalData.dashboard.application.components[0].id)

**Comment by divashdr** [comment]
Thanks, @dardanbekteshi for re-opening this issue!.

**Comment by stevegal** [comment]
it would seem that the widget.js also assumes the name is the type... I'm pretty certain if we took the type of widget through and removes the assumptions that that widget title and the name were the same it would work.

**Comment by dardanbekteshi** [comment]
@divashdr it was not intentional, I closed it by mistake :) 

@stevegal I spent the day investigating the backend and all the methods that do the create/update/delete operations on both the Dashboard and Component classes (and the corresponding MongoDB collections). I think I found the problem, and it mostly lies in the DashboardServiceImpl class. In particular, the methods deleteWidget() and associateCollectorToComponent(). 
But now it is too late, so I will finish with the changes tomorrow morning and will share the code and my observations with you. Then hopefully the needed UI changes will be more clear for you to proceed with.

I am looking forward to collaborating and solving the issue together :D

**Comment by stevegal** [comment]
cool. I was thinking that if we changed the private List<String> activeWidgets; in DashboardRequest to be an object with the title and type we could get it to hold together

**Comment by stevegal** [comment]
@dardanbekteshi yeah... upon making active widgets a List containing title and type, I've chased it bask to the same DashboardServiceImpl class. I can see that the component getCollectorItems is a Map<CollectorType,List<CollectorItems>>.. that would have to change to something that would allow more than one collector type. Just needs one more collection in between...

**Comment by dardanbekteshi** [comment]
@stevegal I made the first changes regarding the delete operation on a widget. You can see the changes here: https://github.com/dardanbekteshi/Hygieia/commit/ea31677c760ee2ed8904aeabe98e8006adaca3c2

I created a branch feature/multiple_widgets which is up to date with the original Hygieia master. I think that is the easiest way to collaborate. Please have a look at the change, maybe checkout the branch and add further changes. 

As it is now, the create and update operations will not work, and the problem I see is in the disable/enable field of a given collector item. Since when we update a given widget, that will mean that we need to disable the previous collector item which was "holding its place" in the component. And the reason for this is that whenever a widget is updated, a new collector item is created in the database, instead of updating the original one. IMO, this makes things more complicated, since as it is right now, we don't have a way to trace back and check which exact collector item is the old one. Because currently only one widget of a given type was supported, this was easy, and we could delete or disable all the collector items of a given type.

My guess is that there has to be a way to send some more information from the frontend to the back, like for example, send all the collector items of the component in a separate parameter, and then send the collector item we are interested in on a separate parameter. Right now, it looks like we are sending a List<CollectorItemId>, but from what I have seen, it is always only one collectoritemid.

But maybe we can further discuss once you have a look at the changes, and we can directly highlight snippets of code we think needs to be changed, might be easier :) 

And also, you can share your UI changes on another branch maybe, so I can also have a look and try to merge backend/frontend changes.

**Comment by dardanbekteshi** [comment]
FYI, if anyone wants to contribute to this feature being developed ASAP you can join me and @stevegal on the following branch: https://github.com/dardanbekteshi/Hygieia/tree/feature/multiple_widgets

ping @HweeWong, @divashdr

New commits, comments, observations are more than welcomed!

**Comment by divashdr** [comment]
@dardanbekteshi @stevegal Thanks

**Comment by stevegal** [comment]
We’ve got multiple widgets configurable on new dashboards, working for build widgets, feature widgets, repo widgets (WIP), code widgets (WIP), performance widgets (WIP), rally widgets (WIP). Looking promising so far. Should I raise a PR before completion for early feedback?

**Comment by dardanbekteshi** [comment]
@stevegal I'd say yes, go ahead!

**Comment by VivekBuzruk** [comment]
Are these changes merged in Master?
